### PR TITLE
Implement getFirstIncompleteBlock and update getCurrentBlockList

### LINKED ISSUE #224: Implement checking blocks for completion
When applicant starts a program application, they are taken to the first block that isn't complete yet.

**Describe the solution you'd like**
Fix the stubbed implementation of `ReadOnlyApplicantProgramServiceImpl#getFirstIncompleteBlock`

**Done when**
Done when starting a new program that already has data for the first N blocks starts on N+1. for the applicant

### PR DESCRIPTION
**Pull request by carolinedanzi**
### Description
1) Adds an `isCompleteWithoutErrors` method to `Block`
2) Implements `getFirstIncompleteBlock` on the read only service - this just gets the first block if it exists from `getCurrentBlockList`
3) Change `getCurrentBlockList` to only provide the list of blocks that are incomplete, have errors, or were completed for this program

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #224 
#406

### CODE REVIEW

File: universal-application-tool-0.0.1/app/services/applicant/Block.java
Comment (kendrickt) [inline universal-application-tool-0.0.1/app/services/applicant/Block.java:61]: (question) if this cannot be memoized, then is it also true that `hasErrors` cannot be memoized also?

related: I think `isComplete` also needs to check `hasErrors`. If all the values are present but there are errors then this block is not complete.

If we ever need an `isFilledOut` method for other reasons, you could make an `isFilledOut` method, which looks like this one, and an `isComplete` that is `isFilledOut() && !hasErrors()`.

**Review by kendrickt — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:66]: Suggestion (blocking): please remove the comment if it is done.

File: universal-application-tool-0.0.1/test/services/applicant/BlockTest.java
Comment (hmhsu) [inline universal-application-tool-0.0.1/test/services/applicant/BlockTest.java:70]: Question (if-minor): can we add more tests for hasErrors? It isn't immediately clear what it does though.

**Review by hmhsu — APPROVED** [review]
Looks good to me.

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (natsid) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:69]: I just filed #406 to track this work. Maybe the best thing would be to attach a TODO to #406 here and/or at `getCurrentBlockList`, and merge this PR for now?

**Review by natsid — APPROVED** [review]
Please see my comment in `ReadOnlyApplicantProgramServiceImpl.java`.

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java]: Thanks for catching that!

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java]: Chatted offline - change to `getCurrentBlockList` is small, so I included it in this PR

File: universal-application-tool-0.0.1/app/services/applicant/Block.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/Block.java]: Good point, I removed memoization from hasErrors and updated `isComplete` to `isCompleteWithoutErrors`

**Review by natsid — APPROVED** [review]
Re-approving after changes - looks good.

File: universal-application-tool-0.0.1/test/services/applicant/BlockTest.java
Comment (kendrickt) [inline universal-application-tool-0.0.1/test/services/applicant/BlockTest.java:134]: a great color!

**Review by kendrickt — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:66]: Thanks! I'm not sure this will work here though - `getCurrentBlockList` returns an `ImmutableList`, not an `Optional`. But I didn't know about map, I'm excited to use it in the future!

File: universal-application-tool-0.0.1/test/services/applicant/BlockTest.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/test/services/applicant/BlockTest.java:69]: That's exciting! Would help tests for sure

File: universal-application-tool-0.0.1/test/services/applicant/BlockTest.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/test/services/applicant/BlockTest.java:134]: :)

File: universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java]: Thank you for that heads up! I moved all the setup logic into the `@Before` method and tests are still failing. I might ask for your help tomorrow

**Comment by carolinedanzi** [comment]
DO NOT MERGE THIS PR

There is a fundamental flaw - if we submit this as-is, the `getCurrentBlockList` computes the list after staging updates, so it thinks the current block is not part of the list shown to applicants. This leads to a problem when we call `getBlockAfter` - the current block we just filled out is no longer in the list, so `getBlockAfter` fails since it can't find the next block to show.

This PR can be merged once we implement metadata checking - if the block is complete but was filled out in this session/program, then it should still be included in the list of current blocks

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java
Comment (bion) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:54]: Suggestion (non-blocking): there's a distinction between the underlying JSON data containing a path and it containing a value at that path i.e. the value at the path could be `null`. Recommend renaming this method to `hasValueAtPath`

Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java]: Good point - updated method name

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (bion) [inline universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:24]: Thought (non-blocking, no recommended change): we don't save updates that result in errors, so we shouldn't ever have a block that is complete but also has errors

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (bion) [inline universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:81]: Suggestion (blocking): as is this would the first block regardless of whether it's incomplete or not I think?

```suggestion
    : getCurrentBlockList().stream().filter(block -> !block.isCompleteWithoutErrors()).findFirst()
```

**Review by bion — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (bion) [inline universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:81]: Suggestion (blocking): as is this would the first block regardless of whether it's incomplete or not I think?

```suggestion
    return getCurrentBlockList().stream().filter(block -> !block.isCompleteWithoutErrors()).findFirst()
```

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:42]: Question (blocking for an answer): if a block is completed, but only partially in this program (some of the questions are answered in other programs), it is not included here. Shall we include them as well?

File: universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java
Comment (hmhsu) [inline universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java:63]: Question (blocking): shouldn't this return questions completed in this program?

File: universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java
Comment (hmhsu) [inline universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java:81]: Question (blocking): should `getCurrentBlockList()` be called again here?

**Review by hmhsu — APPROVED** [review]
Mostly LGTM with a few questions/comments.

File: universal-application-tool-0.0.1/app/services/applicant/Block.java
Comment (bion) [inline universal-application-tool-0.0.1/app/services/applicant/Block.java:89]: Suggestion (blocking): if any of the questions were answered in this program it should be included in this program, otherwise a block could disappear after an applicant completes it?

```suggestion
        .anyMatch(
```

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:81]: Ah I see what  you're saying - this should not return blocks that were completed in this program. I'll update

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:42]: Good catch! Bion caught this too - I'll update `wasCompletedInProgram` to check for any questions in this block that were completed in this program, instead of all questions

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:81]: I would expect that `getCurrentBlockList()` should return blocks completed in this program so applicants can go back?

Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:81]: Agreed - `getCurrentBlockList` should contain the blocks an applicant needs to complete for this program and blocks they have already completed. `getFirstIncompleteBlock` should return the first block the applicant needs to complete, right?

Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:81]: I think my wording in the first response is confusing - it should say "`getFirstIncompleteBlock` should not return blocks that were completed in this program"

Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:81]: Yes! Sorry for the confusion!

File: universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java:63]: Great question - I updated the test name and comment to make it clearer that `getCurrentBlockList` should return empty if all blocks were completed in a different program.

File: universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java:81]: Yes it should! Thank you so much for catching that!

File: universal-application-tool-0.0.1/app/services/applicant/Block.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/Block.java:89]: Thanks! Updated

File: universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:24]: Good point that we don't save error blocks - I'm wondering if this might come into play during updates still. We build a `ReadOnlyApplicantProgramService` based on the current block we're updating and the ApplicantData where we just staged updates, so that (I think) we can determine if the next block for the applicant is still the block they were working on if it has errors

**Pull request merged**
